### DISCUSSION TITLE
GoogleMapsOverlay error: texImage2D: no bound PIXEL_UNPACK_BUFFER

### DISCUSSION THREAD

**Discussion by mantox74** [visgl/deck.gl#6510] [Q&A]
Hallo everyone,
I'am new on deck.gl.
I'm trying to create a google map in Angular with about 380k of markers.
I have created an IconLayer

```
import { GoogleMapsOverlay as DeckOverlay } from '@deck.gl/google-maps';
........
this.pointOverlay = new DeckOverlay();
this.renderLayer();
......
renderLayer() {
    this.pointOverlay.finalize();
    this.pointOverlay.setProps({
      layers: [
        new IconLayer({
          id: 'dati_geojson_icon',
          data: this.jsonData,
          getPosition: (d) => d.geometry.coordinates,
          getIcon: (d) => 'marker',
          iconAtlas: 'https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/icon-atlas.png',
          iconMapping: {
            marker: {
              x: 128,
              y: 0,
              width: 128,
              height: 128,
            },
          },
          getSize: (d) => 5,
          sizeScale: 8,
          onClick: (info) =>
            info.object &&
            console.log('..'),
        }),
      ],
    });

    this.heatmapOverlay = new DeckOverlay({
      layers: [
        new HeatmapLayer({
          id: 'heatmapLayer',
          data: this.jsonData,
          getPosition: (d) => d.geometry.coordinates,
          getWeight: (d) => 1,
          aggregation: 'SUM',
        }),
      ],
    });

    // show default layer
    this.pointOverlay.setMap(this.map);
  }
```

I'm trying to use a demo example to integrate my iconAtlas and map one marker but when I run my project and show the google map I have these console warning and black squares on map (no markers image):

1. **WebGL: INVALID_OPERATION: texImage2D: no bound PIXEL_UNPACK_BUFFER**
2. **[.WebGL-0x7fd54f950600]GL ERROR :GL_INVALID_OPERATION : glGenerateMipmap: Can not generate mips**

And then I have also this warning every time I move mouse on square.
**[.WebGL-0x7fd54f950600]RENDER WARNING: texture bound to texture unit 0 is not renderable. It might be non-power-of-2 or have incompatible texture filtering (maybe)?**

I have searched on web what should be the problem but I can't find nothing. Also in documentation there is not any (or I can't find it).

Can any one help me, suggest a solution?

**Comment by Pessimistress** [comment]
You can't call `setProps` after `finalize`. `finalize` disposes all resources and as the name suggests, is final.

**Reply by mantox74** [reply]
Thank's for your answer but it works with getIcon, the problem is only when I try to use iconAtlas
This works:
```
.....
getIcon: (d) => ({
   url: 'assets/img/pin.png',
   width: 30,
   height: 40,
   anchorY: 30,
}),
.....
```

**Comment by massimiliamanto74** [comment]
Any suggestions?

**Reply by Pessimistress** [reply]
You may want to check out this issue: https://github.com/visgl/deck.gl/issues/4598

**Reply by massimiliamanto74** [reply]
Hi @Pessimistress,
thank's for your response.
I saw this issue some days ago but if i try to add this:

```
import { registerLoaders } from '@loaders.gl/core';
import { ImageLoader } from '@loaders.gl/images';
constructor(){
registerLoaders([ImageLoader]); 
}
```

I receive many errors:

```
Error: node_modules/@loaders.gl/images/src/lib/utils/globals.ts:43:32 - error TS2304: Cannot find name 'importScripts'.

43 export const isWorker = typeof importScripts === 'function';
                                  ~~~~~~~~~~~~~


Error: node_modules/@loaders.gl/loader-utils/src/lib/env-utils/globals.ts:26:41 - error TS2304: Cannot find name 'importScripts'.

26 export const isWorker: boolean = typeof importScripts === 'function';
                                           ~~~~~~~~~~~~~


Error: node_modules/@loaders.gl/loader-utils/src/lib/node/fs.ts:2:8 - error TS1192: Module '"fs"' has no default export.

2 import fs from 'fs';
         ~~


Error: node_modules/@loaders.gl/worker-utils/src/lib/env-utils/globals.ts:25:41 - error TS2304: Cannot find name 'importScripts'.

25 export const isWorker: boolean = typeof importScripts === 'function';
                                           ~~~~~~~~~~~~~


Error: node_modules/@loaders.gl/worker-utils/src/lib/library-utils/library-utils.ts:86:12 - error TS2304: Cannot find name 'importScripts'.

86     return importScripts(libraryUrl);
              ~~~~~~~~~~~~~


Error: node_modules/@loaders.gl/worker-utils/src/lib/process-utils/process-utils.ts:1:8 - error TS2613: Module '"child_process"' has no default export. Did you mean to use 'import { ChildProcess } from "child_process"' instead?

1 import ChildProcess from 'child_process';
         ~~~~~~~~~~~~


Error: node_modules/apache-arrow/recordbatch.d.ts:17:18 - error TS2430: Interface 'RecordBatch<T>' incorrectly extends interface 'StructVector<T>'.
  The types of 'slice(...).clone' are incompatible between these types.
    Type '(data: Data<Struct<T>>, children?: AbstractVector<any>[]) => RecordBatch<T>' is not assignable to type '<R extends DataType<Type, any> = Struct<T>>(data: Data<R>, children?: AbstractVector<R>[]) => VectorType<R>'.
      Types of parameters 'data' and 'data' are incompatible.
        Type 'Data<R>' is not assignable to type 'Data<Struct<T>>'.
          Type 'R' is not assignable to type 'Struct<T>'.
            Property 'dataTypes' is missing in type 'DataType<Type, any>' but required in type 'Struct<T>'.

17 export interface RecordBatch<T extends {
                    ~~~~~~~~~~~

  node_modules/apache-arrow/type.d.ts:457:5[0m
    457     dataTypes: T;
            ~~~~~~~~~
    'dataTypes' is declared here.

Error: node_modules/apache-arrow/recordbatch.d.ts:24:22 - error TS2415: Class 'RecordBatch<T>' incorrectly extends base class 'StructVector<T>'.

24 export declare class RecordBatch<T extends {
```

I don't thinks it's the right way, also because the error in issue is different form mine.

I have tried to change the format of the image but I have the same problem (black sqaures instead icon sprite as you can see under the map) as can you see in screenshot:

<img width="614" alt="Schermata 2021-12-16 alle 14 55 30" src="https://user-images.githubusercontent.com/62379864/146385113-389c4a68-e52f-41f7-b4b4-63375b399c38.png">


This is the error I receive

<img width="592" alt="Schermata 2021-12-16 alle 14 55 36" src="https://user-images.githubusercontent.com/62379864/146385295-f173f1ed-e010-4e6c-9285-154ddb6b0181.png">


The code in file **node_modules/@luma.gl/webgl/dist/esm/classes/texture.js**
<img width="698" alt="Schermata 2021-12-16 alle 15 08 20" src="https://user-images.githubusercontent.com/62379864/146387128-d80e0aab-a832-40af-bb01-47d86bcea186.png">


I am really frusted!

Can someone help me?

**Reply by Pessimistress** [reply]
To check if you are having the same issue as #4598, put a break point where the WebGL error is (`gl.texImage2D(...)`) and inspect `data`. `data` should be an `ImageBitmap` if the image is loaded correctly.

Image format would not make a difference.

You can set `skipLibChecks: true` in your tsconfig to remove those errors.

**Reply by massimiliamanto74** [reply]
Hi @Pessimistress,
thanks a lot for yours responses.
I have inspected **data** as you suggested but it is not a ImageBitmap but it is a Promise in pending status.
I am investigating the problem but it still exists.

**Reply by Pessimistress** [reply]
That is strange. Can you put a break point [here](https://github.com/visgl/deck.gl/blob/v8.6.4/modules/core/src/lifecycle/component-state.js#L152) and see if the promise fails the `instanceof Promise` test?

**Reply by massimiliamanto74** [reply]
@Pessimistress yes the value is an Array of features (GeoJson format) so the instanceof Promise check fails and do not run the code
```this._watchPromise(propName, value);```

**Reply by Pessimistress** [reply]
You will hit that break point multiple times. You are looking for the time when `propName` is `iconAtlas`.

**Comment by ibgreen** [comment]
@massimiliamanto74 The typescript issues you are seeing were fixed in loaders.gl 3.1.0

**Reply by massimiliamanto74** [reply]
Hi @ibgreen,
thanks for you response but the problem is that the image atlas is not loading, the warning typescript is secondary.

Thank's
